src/hyperscan.rs

```diff
@@ -0,0 +1,299 @@
+        (P_TEST, "test".to_string()),
+        (P_EMAIL, "([\\w\\.!#$%&'*+\\-=?\\^_`{|}~]+@([\\w\\d-]+\\.)+[\\w]{2,4})".to_string()),
+        (P_IP, "(((25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9][0-9]|[0-9])\\.){3}(25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9][0-9]|[0-9]))".to_string()),
+        (P_QUOTED, "((\"(.*?)\")|'(.*?)')".to_string()),
+        (P_DATE, "((19[789]\\d|2\\d{3})[-/](0[1-9]|1[1-2])[-/](0[1-9]|[1-2][0-9]|3[0-1]*))|((0[1-9]|[1-2][0-9]|3[0-1]*)[-/](Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|(0[1-9]|1[1-2]))[-/](19[789]\\d|2\\d{3}))".to_string()),
+        (P_PHONE, "[\\(]?(\\d{3})[\\)-]?[- ]?(\\d{3})[- ]?(\\d{4})".to_string()),
+        (P_USER_AGENT, "\"((Mozilla|Links).*? \\(.*?\\)( .*?[0-9]{1,3}\\.[0-9]{1,3}\\.?[0-9]{0,3})?)\"".to_string()),
+        (P_URL, "(https?|ftp)://[^\\s/$.?#].[^()\\]\\[\\s]*".to_string()),
+    ].iter().cloned().collect();
+
+    let mut patterns: Vec<Pattern> = Vec::new();
+
+    if flags.contains(constants::ScanFlags::IP) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_IP).unwrap().clone(),
+            id: P_IP.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::EMAIL) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_EMAIL).unwrap().clone(),
+            id: P_EMAIL.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::DATE) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_DATE).unwrap().clone(),
+            id: P_DATE.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::QUOTED) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_QUOTED).unwrap().clone(),
+            id: P_QUOTED.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::URL) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_URL).unwrap().clone(),
+            id: P_URL.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::PHONE) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_PHONE).unwrap().clone(),
+            id: P_PHONE.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::USER_AGENT) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_USER_AGENT).unwrap().clone(),
+            id: P_USER_AGENT.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+
+    let res_db: Result<BlockDatabase, _> = patterns.build();
```

Comment:
We might want to keep the database around, there is serialization & deserialization support: http://intel.github.io/hyperscan/dev-reference/serialization.html

Reply:
We keep this DB around only for as long as this query is, but we could definitively serialize and keep it, we can experiment with that.